Libraries/LibFTP/FTPConnection.cpp

```diff
@@ -0,0 +1,188 @@
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "FTPConnection.h"
+
+#include <LibCore/TCPSocket.h>
+
+namespace FTP {
+FTPConnection::FTPConnection()
+    : m_socket(Core::TCPSocket::construct())
+{
+}
+FTPConnection::~FTPConnection()
+{
+}
+
+void FTPConnection::connect(const URL& url, Function<void()> callback)
+{
+    auto address = url.host();
+    auto port = url.port();
+    m_socket->on_connected = [&]() {
+        // FIXME: auto login and set path
+        callback();
+    };
+    m_socket->connect(address, port);
+}
+void FTPConnection::disconnect()
+{
+    m_socket->close();
+}
+void FTPConnection::set_transfer_mode(TransferMode mode)
+{
+    ASSERT_NOT_REACHED();
+    return;
+
+    m_transfer_mode = mode;
+    // FIXME: error handling, mode to FTP code
+
+    FTPRequest request;
+    request.set_command(FTPRequest::TransferMode);
+    //request.add_arg(mode);
+    auto response = send_command(request);
+}
+void FTPConnection::set_data_representation(DataRepresentation data_representation)
+{
+    ASSERT_NOT_REACHED();
+    return;
+
+    m_data_representation = data_representation;
+    // FIXME: error handling, mode to FTP code
+
+    FTPRequest request;
+    request.set_command(FTPRequest::TransferType);
+    //request.add_arg(data_representation);
+    auto response = send_command(request);
+}
+void FTPConnection::login(const String& login, const String& password)
+{
+    // FIXME: error handling
+
+    FTPRequest request;
+    request.set_command(FTPRequest::Username);
+    request.add_arg(login);
+    auto response = send_command(request);
+
+    request.set_command(FTPRequest::Password);
+    request.add_arg(password);
+    response = send_command(request);
+}
+void FTPConnection::set_remote_directory(const String& remote_dir)
+{
+    FTPRequest request;
+    request.set_command(FTPRequest::PrintWorkingDirectory);
```

Comment:
This function is called `set_remote_directory` yet it calls `PrintWorkingDirectory`, which sends `PWD`.